crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+    let error = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let other_len = decoder.read_u16()?.unverified(/*valid as any u16*/);
+    let other =
+        decoder.read_vec(other_len as usize)?.unverified(/*valid as any vec ot the right size*/);
+
+    Ok(TSIG {
+        algorithm,
+        time,
+        fudge,
+        mac,
+        oid,
+        error,
+        other,
+    })
+}
+
+/// Write the RData from the given Encoder
+///
+/// ```text
+///                       1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+///   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  /                         Algorithm Name                        /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                                                               |
+///  |          Time Signed          +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                               |            Fudge              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          MAC Size             |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+             MAC               /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Original ID          |            Error              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Other Len            |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+           Other Data          /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+pub fn emit(encoder: &mut BinEncoder<'_>, tsig: &TSIG) -> ProtoResult<()> {
+    tsig.algorithm.emit(encoder)?;
+    encoder.emit_u16((tsig.time >> 32) as u16)?;
+    encoder.emit_u32(tsig.time as u32)?;
+    encoder.emit_u16(tsig.fudge)?;
+    encoder.emit_u16(tsig.mac.len() as u16)?;
+    encoder.emit_vec(&tsig.mac)?;
+    encoder.emit_u16(tsig.oid)?;
+    encoder.emit_u16(tsig.error)?;
+    encoder.emit_u16(tsig.other.len() as u16)?;
+    encoder.emit_vec(&tsig.other)?;
+    Ok(())
+}
+
+// Does not appear to have a normalized text representation
+impl fmt::Display for TSIG {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        write!(
+            f,
+            "{algorithm} {time} {fudge} {mac} {oid} {error} {other}",
+            algorithm = self.algorithm,
+            time = self.time,
+            fudge = self.fudge,
+            mac = sshfp::HEX.encode(&self.mac),
+            oid = self.oid,
+            error = self.error,
+            other = sshfp::HEX.encode(&self.other),
+        )
+    }
+}
+
+impl Algorithm {
+    /// Return DNS name for the algorithm
+    pub fn to_name(&self) -> Name {
```

Comment:
I didn't read the entire RFC, this is intended to be a DNS label?

Reply:
From https://datatracker.ietf.org/doc/html/rfc8945#section-4.2
> Algorithm Name:
      an octet sequence identifying the TSIG algorithm in the domain
      name syntax.  (Allowed names are listed in Table 3.)  **The name is
      stored in the DNS name wire format** as described in [RFC1034].  As
      per [RFC3597], this name MUST NOT be compressed.

Reply:
Which makes me realize it can be compressed with current implementation, it could in theory be compressed. It is usually not because you can't have more than a single tsig, so it would require querying a domain which collide with an algorithm name, but I'll still going to fix that